File: 백준/Silver/24445. 알고리즘 수업 － 너비 우선 탐색 2/알고리즘 수업 － 너비 우선 탐색 2.py
import sys
from collections import deque

def bfs():
  global visited, q, count
  while q:
    cur = q.popleft()
    for next in graph[cur]:
      if not visited[next]:
        count+=1
        visited[next] = count
        q.append(next)
        
    

V, E, start = map(int, sys.stdin.readline().strip().split())
graph = [[] for _ in range(V+1)] 
visited = [0] * (V+1) # 방문 순서 확인을 위해서 만듦
q = deque() # bfs 큐
count = 1 # 몇 번째 순서로 들어가는지 

# 인접리스트를 사용하기 때문에 
for i in range(E):
  a,b = map(int, sys.stdin.readline().strip().split())
  graph[a].append(b)
  graph[b].append(a)
# 내림차순으로 방문을 해서 그래프를 내림차순으로 정렬
for i in range(V+1):
  graph[i].sort(reverse=True)

q.append(start) # 큐에 시작 정점을 먼저 넣음
visited[start] = count # 1번째로 방문했으니 체크
bfs()

for i in range(1, len(visited)): # 1번부터 순서를 뽑아냄.
  print(visited[i])

File: 백준/Silver/24445. 알고리즘 수업 － 너비 우선 탐색 2/알고리즘 수업 － 너비 우선 탐색 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;


public class Main {

    static int N;
    static int M;
    static int R;
    static boolean[] visited;
    static List<Integer>[] graph;
    static int[] result;
    static int count = 1;

    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        String[] s = bf.readLine().split(" ");
        N = Integer.parseInt(s[0]);
        M = Integer.parseInt(s[1]);
        R = Integer.parseInt(s[2]);

        graph = new List[N+1];
        visited = new boolean[N+1];
        result = new int[N+1];

        for(int i=0; i<graph.length; i++){
            graph[i] = new ArrayList<>();
        }



        for(int i=0; i<M; i++){
            s = bf.readLine().split(" ");
            int start = Integer.parseInt(s[0]);
            int end = Integer.parseInt(s[1]);

            graph[start].add(end);
            graph[end].add(start);
        }

        for(int i=1; i<graph.length; i++){
            Collections.sort(graph[i], Collections.reverseOrder());
        }
        bfs();

        for(int i=1; i<result.length; i++){
            System.out.println(result[i]);
        }


    }

    public static void bfs(){
        Queue<Integer> q = new LinkedList<>();
        q.add(R);
        visited[R] = true;
        result[R] = count++;

        while(!q.isEmpty()){
            int cur = q.poll();
            for(int i=0; i<graph[cur].size(); i++){
                int next = graph[cur].get(i);
                if(!visited[next]){
                    visited[next] = true;
                    result[next] = count++;
                    q.add(next);
                }
            }
        }
    }
}